manual_hook.rb

```diff
@@ -2,56 +2,75 @@
 
 require 'resolv'
 
-def setup_dns(domain, txt_challenge)
-  resolved = false;
-  singleLoop = false;
-  dns = Resolv::DNS.new;
-  acme_domain = "_acme-challenge."+domain; 
-  puts "Checking for pre-existing TXT record for the domain: \'#{acme_domain}\'."
-
-  until resolved
-    dns.each_resource(acme_domain, Resolv::DNS::Resource::IN::TXT) { |resp|
-     if resp.strings[0] == txt_challenge
-       puts "Found #{resp.strings[0]}. match."
-       resolved = true;
-     else
-       puts "Found #{resp.strings[0]}. no match."
-     end
-    }
-
-    if !resolved
-     if !singleLoop
-       puts "Create TXT record for the domain: \'#{acme_domain}\'. TXT record:"
-       puts "\'#{txt_challenge}\'"
-       puts "Press enter when DNS has been updated..."
-       $stdin.readline()
-       singleLoop = true
-     end
-
-     puts "Didn't find a match for #{txt_challenge}"; 
-     puts "Waiting to retry..."; 
-     sleep 30; 
+# Struct to store a challenge
+Challenge = Struct.new(:domain, :acme_domain, :txt_challenge)
+
+
+# Check if a challenge is resolved, returns true in that case
+def resolved?(dns, challenge)
+  dns.each_resource(challenge[:acme_domain], Resolv::DNS::Resource::IN::TXT) { |resp|
+    if resp.strings[0] == challenge[:txt_challenge]
+      puts "Found #{resp.strings[0]}. match."
+      return true
+    else
+      puts "Found #{resp.strings[0]}. no match."
+      return false
+    end
+  }
+  puts "Didn't find a match for #{challenge[txt_challenge]}"
+  return false
+end
+
+def setup_dns(challenges)
+  # DNS Resolver
+  dns = Resolv::DNS.new
+  first_iteration = true
+  until challenges.empty? # Until all challenges are resolved
+    challenges.delete_if do |challenge| # Delete resolved challenges
+      puts "Checking for TXT record for the domain: \'#{challenge[:acme_domain]}\'."
+      resolved?(dns, challenge)
+    end
+
+    if first_iteration
+      challenges.each do |challenge|
+        puts "Create TXT record for the domain: \'#{challenge[:acme_domain]}\'. TXT record:"
+        puts "\'#{challenge[:txt_challenge]}\'"
+      end
+      puts "Press enter when DNS has been updated..."
+      $stdin.readline
+      first_iteration = false
+    else
```

Comment:
I always waited till i saw the DNS records propagated (i have set a TTL of 300s for them) and then pressed enter. In that case the 2 minutes were unnecessary for me.

I can revert the change if you like.

Reply:
ohhh, interesting. i let the computer™ confirm it for me, because looking at DNS records is like watching :grass: grow. 😉 At the end of the day, the same thing happens, and you have a valid use case.

👍